gitlab/resource_gitlab_topic.go

```diff
@@ -0,0 +1,153 @@
+			"soft_destroy": {
+				Description: "Empty the topics fields instead of deleting it",
+				Type:        schema.TypeBool,
+				Required:    true,
+			},
+			"description": {
+				Description: "A text describing the topic",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+		},
+	}
+}
+
+func resourceGitlabTopicCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	options := &gitlab.CreateTopicOptions{
+		Name: gitlab.String(d.Get("name").(string)),
+	}
+
+	if v, ok := d.GetOk("description"); ok {
+		options.Description = gitlab.String(v.(string))
+	}
+
+	log.Printf("[DEBUG] create gitlab topic %s", *options.Name)
+
+	topic, _, err := client.Topics.CreateTopic(options, gitlab.WithContext(ctx))
+	if err != nil {
+		return diag.Errorf("Failed to create topic %q: %s", *options.Name, err)
+	}
+
+	d.SetId(fmt.Sprintf("%d", topic.ID))
+
+	return resourceGitlabTopicRead(ctx, d, meta)
+}
+
+func resourceGitlabTopicRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+
+	topicID, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return diag.Errorf("Failed to convert topic id %s to int: %s", d.Id(), err)
+	}
+	log.Printf("[DEBUG] read gitlab topic %d", topicID)
+
+	topic, _, err := client.Topics.GetTopic(topicID, gitlab.WithContext(ctx))
+	if err != nil {
+		if is404(err) {
+			log.Printf("[DEBUG] gitlab group %s not found so removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return diag.Errorf("Failed to read topic %d: %s", topicID, err)
+	}
+
+	d.SetId(fmt.Sprintf("%d", topic.ID))
+	d.Set("name", topic.Name)
+	d.Set("description", topic.Description)
+
+	return nil
+}
+
+func resourceGitlabTopicUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	options := &gitlab.UpdateTopicOptions{}
+
+	if d.HasChange("name") {
+		options.Name = gitlab.String(d.Get("name").(string))
+	}
+
+	if d.HasChange("description") {
+		options.Description = gitlab.String(d.Get("description").(string))
+	}
+
+	log.Printf("[DEBUG] update gitlab topic %s", d.Id())
+
+	topicID, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return diag.Errorf("Failed to convert topic id %s to int: %s", d.Id(), err)
+	}
+	_, _, err = client.Topics.UpdateTopic(topicID, options, gitlab.WithContext(ctx))
+	if err != nil {
+		return diag.Errorf("Failed to update topic %d: %s", topicID, err)
+	}
+
+	return resourceGitlabTopicRead(ctx, d, meta)
+}
+
+func resourceGitlabTopicDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+
+	softDestroy := d.Get("soft_destroy").(bool)
+	warning := fmt.Sprintf("[WARN] Not deleting gitlab topic %s as gitlab API doens't support deleting topics", d.Id())
+
+	if softDestroy {
+		warning += ". Instead emptying its description"
+	}
+
+	log.Println(warning)
+
+	if !softDestroy {
+		return nil
```

Comment:
This should return an error saying something like "Destroying a topic is not yet supported. You can set soft_destroy=true to suppress this error."